packages/firestore/test/unit/specs/listen_spec.test.ts

```diff
@@ -785,4 +785,62 @@ describeSpec('Listens:', [], () => {
       .client(2)
       .expectEvents(query, { added: [docB] });
   });
+
+  specTest('Query recovers after primary takeover', ['multi-client'], () => {
+    const query = Query.atPath(path('collection'));
+    const docA = doc('collection/a', 1000, { key: 'a' });
+    const docB = doc('collection/b', 2000, { key: 'b' });
+
+    return (
+      client(0)
+        .expectPrimaryState(true)
+        .userListens(query)
+        .watchAcksFull(query, 1000, docA)
+        .expectEvents(query, { added: [docA] })
+        .client(1)
+        .userListens(query)
+        .expectEvents(query, { added: [docA] })
+        .stealPrimaryLease()
+        .expectListen(query, 'resume-token-1000')
+        .watchAcksFull(query, 2000, docB)
+        .expectEvents(query, { added: [docB] })
+        .client(0)
+        .expectPrimaryState(false)
+        // TODO(multitab): Suppres the additional `fromCache`, which is raised
+        // locally because we disable the network when we lose the primary
+        // lease.
```

Comment:
FWIW I'm not sure if this is wrong / harmful.  When a primary switch happens we collectively will be without a network connection for a period of time and it seems correct for views to trigger fromCache=true events...  but we _should_ return to fromCache=false once the primary switch completes and a new primary establishes the listens...